electron/new_config.ts

```diff
@@ -0,0 +1,294 @@
+import { exec } from 'child_process'
+import {
+  existsSync,
+  mkdir,
+  readFileSync,
+  readdirSync,
+  writeFileSync
+} from 'graceful-fs'
+import { userInfo as user } from 'os'
+
+import { AppSettings, ConfigVersion, WineProps } from './types'
+import {
+  currentConfigVersion,
+  heroicConfigPath,
+  heroicGamesConfigPath,
+  heroicInstallPath,
+  heroicToolsPath,
+  home,
+  userInfo
+} from './constants'
+import {
+  execAsync
+} from './utils'
+
+/**
+ * This class does config handling.
+ * This can't be constructed directly. Use the static method get().
+ * It automatically selects the appropriate config loader based on the config version.
+ *
+ * It also implements all the config features that won't change across versions.
+ */
+abstract class GlobalConfig {
+  protected static globalInstance : GlobalConfig
+
+  public abstract version : ConfigVersion
+
+  public config : AppSettings
+
+  /**
+   * Get the global configuartion handler.
+   *
+   * @param version ConfigVersion to use for parsing config file. Default: 'auto'
+   * @returns GlobalConfig instance or undefined.
+   */
+  public static get(version : ConfigVersion = 'auto') : GlobalConfig {
+    let configVersion : ConfigVersion = undefined
+
+    // Config file doesn't already exist, make one with the current version.
+    if (!existsSync(heroicConfigPath)) {
+      configVersion = currentConfigVersion
+
+      // For autodetect.
+      if (version === 'auto') {
+        version = currentConfigVersion
+      }
+    }
+    // Config file exists, detect its version.
+    else {
+      // Check version field in the config.
+      configVersion = JSON.parse(readFileSync(heroicConfigPath, 'utf-8'))['version']
+      // Legacy config file without a version field, it's a v0 config.
+      if (configVersion === undefined) {
+        configVersion = 'v0'
+      }
+
+      // For autodetect.
+      if (version === 'auto') {
+        version = configVersion
+      }
+    }
+
+    // Check for version mismatches.
+    // Should never happen for 'auto'.
+    if (version !== configVersion) {
+      console.log(`Config version mismatch! Requested: ${version}. Found: ${configVersion}.`)
+    }
+
+    // Select loader to use.
+    switch (version) {
+    case 'v0':
+      if (GlobalConfig.globalInstance === undefined) {
```

Comment:
same here, you can remove the equals undefined. 
The thing here is that if GlobalConfig for some reason does not exist, you will have an error like `there is no globalnstance in undefined` or something like that.
If GlobalConfig can be undefined, try doing something like this:
`if (GlobalConfig?.globalInstance) {`
Then it will run if both are undefined as well.

Reply:
GlobalConfig is always defined. Shouldn't be an issue. I'll remove the unnecessary `undefined` parts.